File: 03월 25일 - 백트래킹/과제/10971.py
import sys
input = sys.stdin.readline

"""
[외판원 순회2]
모든 도시를 방문할 수 있는 사이클을 만들고, 그 중 최소비용을 구한다.
사이클이 형성되므로 출발 도시는 중요하지 않다 -> 0으로 고정
"""

MAX = 10**8

# curr_city: 현재 도시의 index, left: 남은 도시의 수, cost: 현재까지의 경비
def backtracking(count, curr_city, cost):
    global answer   # 전역변수 answer 업데이트 필요
    
    if cost > answer:   # 생각해보기 : 이 조건문이 없으면 어떻게 될까요?
        return

    if count == n - 1:   # 모든 도시를 다 방문했다면 0(출발도시)으로 돌아올 수 있는지 확인
        if matrix[curr_city][0] > 0:
            answer = min(answer, cost + matrix[curr_city][0])
        return

    for next_city in range(n):
        if visited[next_city] or matrix[curr_city][next_city] == 0:
            continue
        visited[next_city] = True
        backtracking(count + 1, next_city, cost + matrix[curr_city][next_city])
        visited[next_city] = False

    return


# 입력
n = int(input())
matrix = [list(map(int, input().split())) for _ in range(n)]

visited = [False] * n
answer = MAX

visited[0] = True
backtracking(0, 0, 0)   # 0에서 출발해 n - 1개 도시를 방문
print(answer)

File: 03월 25일 - 백트래킹/과제/10971.cpp
#include <iostream>
#include <vector>

using namespace std;
const int INF = 1e8;

vector<bool> visited; //방문 검사 배열
vector<vector<int>> matrix; //비용 행렬
int n, ans = INF; //최소 비용 저장 위해 미리 최댓값으로 초기화

/**
 * !생각해보기! 함수의 각 인자는 각각 어떤 역할을 하고 있나요?
 */
void backtracking(int cnt, int cur_city, int cost) {
    if (cost >= ans) { //생각해보기 : 이 조건문이 없으면 어떻게 될까요?
        return;
    }
    if (cnt == n) { //종료 조건 : n개의 도시를 순회했음
        //출발 도시(0)로 다시 돌아올 수 있는지 확인
        if (matrix[cur_city][0] != 0) {
            ans = min(ans, cost + matrix[cur_city][0]);
        }
        return;
    }
    for (int i = 0; i < n; i++) { //cur_city에서 도시 i로 이동
        if (matrix[cur_city][i] && !visited[i]) { //길이 있고, 아직 방문하지 않은 도시
            visited[i] = true;
            backtracking(cnt + 1, i, cost + matrix[cur_city][i]);
            visited[i] = false;
        }
    }
}

/**
 * 백트래킹 풀이 (4ms)
 *
 * 한 사이클을 만들면, 어느 도시에서 시작하든 값은 같으므로 0번 도시부터 시작하는 경우만 검사하면 된다.
 */

int main() {
    //입력
    cin >> n;
    visited.assign(n, false);
    matrix.assign(n, vector<int>(n, 0));
    for (int i = 0; i < n; i++) {
        for (int j = 0; j < n; j++) {
            cin >> matrix[i][j];
        }
    }

    //연산
    visited[0] = true;
    backtracking(1, 0, 0);

    //출력
    cout << ans;
    return 0;
}